geomstats/learning/em_expectation_maximization.py

```diff
@@ -0,0 +1,303 @@
+        """
+        self.mixture_coefficients = gs.mean(posterior_probabilities, 0)
+
+    def update_means(self, data, posterior_probabilities,
+                     lr_means, tau_means, max_iter=DEFAULT_MAX_ITER):
+        """Means update function."""
+        n_gaussian = posterior_probabilities.shape[-1]
+
+        mean = FrechetMean(
+            metric=self.riemannian_metric,
+            method=self.mean_method,
+            lr=lr_means,
+            tau=tau_means,
+            max_iter=max_iter,
+            point_type=self.point_type)
+
+        data_expand = gs.expand_dims(data, 1)
+        data_expand = gs.repeat(data_expand, n_gaussian, axis=1)
+
+        mean.fit(data_expand, weights=posterior_probabilities)
+        self.means = gs.squeeze(mean.estimate_)
+
+    def update_variances(self, data, posterior_probabilities):
+        """Variances update function.
+
+        Parameters
+        ----------
+        data : array-like, shape=[n_samples, n_features]
+            Training data, where n_samples is the number of samples and
+            n_features is the number of features.
+        posterior_probabilities : array-like, shape=[n_samples, n_gaussian]
+            Probability of a given sample to belong to a component
+            of the GMM, computed for all components.
+        """
+        n_data, n_gaussian = data.shape[0], self.means.shape[0]
+
+        data_expand = gs.expand_dims(data, 1)
+        data_expand = gs.repeat(data_expand, n_gaussian, axis=1)
+        means = gs.expand_dims(self.means, 0)
+        means = gs.repeat(means, n_data, axis=0)
+
+        data_flattened = gs.reshape(data_expand, (-1, data_expand.shape[-1]))
+        means_flattened = gs.reshape(means, (-1, means.shape[-1]))
+
+        dist_means_data = (self.riemannian_metric.dist(
+            data_flattened, means_flattened) ** 2)
+
+        dist_means_data = gs.reshape(dist_means_data,
+                                     (data_expand.shape[0],
+                                      data_expand.shape[1]))
+
+        weighted_dist_means_data = (dist_means_data *
+                                    posterior_probabilities).sum(0) / \
+            posterior_probabilities.sum(0)
+
+        self.variances = \
+            self.riemannian_metric.find_variance_from_index(
+                weighted_dist_means_data)
+
+    def _expectation(self, data):
+        """Update the posterior probabilities.
+
+        Parameters
+        ----------
+        data : array-like, shape=[n_samples, n_features]
+            Training data, where n_samples is the number of samples and
+            n_features is the number of features.
+        """
+        probability_distribution_function = \
+            PoincareBall.pdf(
+                data, self.means, self.variances,
+                norm_func=self.riemannian_metric.find_normalization_factor,
+                metric=self.riemannian_metric)
+
+        if (probability_distribution_function.mean() !=
```

Comment:
Is this equivalent to `np.isnan`? or `np.isnan` or `is None`?

Reply:
yes, replaced with gs.isnan